File: src/tcp_echo.zig
const std = @import("std");
const IoBackend = @import("utils.zig").IoBackend;

// Golden TCP benchmark: `--conns` concurrent loopback connections, each doing
// `--msgs` request/response round-trips of `--size` bytes against a
// per-connection echo handler. Two golden presets:
//
//   defaults (1000 conns x 100 msgs x 64B)   many-connection throughput
//   --conns=1 --msgs=100000 --size=4096      single-connection latency chain
//                                            (the old tcp_ping_pong)
//
// Counterparts: go/tcp_echo, rust tcp_echo (tokio), cpp/tcp_echo_photon.
const port: u16 = 18766;
const max_msg_size = 4096;

const Config = struct {
    conns: u64 = 1000,
    msgs: u64 = 100,
    size: u64 = 64,
};

fn echoHandler(io: std.Io, stream: std.Io.net.Stream, size: u64) void {
    defer stream.close(io);
    var read_buf: [max_msg_size]u8 = undefined;
    var write_buf: [max_msg_size]u8 = undefined;
    var msg: [max_msg_size]u8 = undefined;
    var r = stream.reader(io, read_buf[0..size]);
    var w = stream.writer(io, write_buf[0..size]);
    while (true) {
        r.interface.readSliceAll(msg[0..size]) catch return;
        w.interface.writeAll(msg[0..size]) catch return;
        w.interface.flush() catch return;
    }
}

fn serverTask(io: std.Io, cfg: Config, ready: *std.Io.Event) void {
    const addr: std.Io.net.IpAddress = .{ .ip4 = std.Io.net.Ip4Address.loopback(port) };
    var srv = addr.listen(io, .{ .reuse_address = true, .kernel_backlog = @intCast(@min(cfg.conns, 4096)) }) catch |err| {
        std.log.err("server: listen failed: {}", .{err});
        ready.set(io);
        return;
    };
    defer srv.deinit(io);

    ready.set(io);

    // One echo handler per accepted connection; wait for all before returning.
    var handlers: std.Io.Group = .init;
    defer handlers.cancel(io);

    var i: u64 = 0;
    while (i < cfg.conns) : (i += 1) {
        const stream = srv.accept(io) catch |err| {
            std.log.err("server: accept failed: {}", .{err});
            break;
        };
        handlers.concurrent(io, echoHandler, .{ io, stream, cfg.size }) catch {
            stream.close(io);
            break;
        };
    }
    handlers.await(io) catch {};
}

fn clientTask(io: std.Io, cfg: Config, ready: *std.Io.Event) void {
    ready.wait(io) catch return;

    const addr: std.Io.net.IpAddress = .{ .ip4 = std.Io.net.Ip4Address.loopback(port) };
    var stream = addr.connect(io, .{ .mode = .stream }) catch |err| {
        std.log.err("client: connect failed: {}", .{err});
        return;
    };
    defer stream.close(io);

    var read_buf: [max_msg_size]u8 = undefined;
    var write_buf: [max_msg_size]u8 = undefined;
    var msg: [max_msg_size]u8 = @splat(0);
    var r = stream.reader(io, read_buf[0..cfg.size]);
    var w = stream.writer(io, write_buf[0..cfg.size]);

    var i: u64 = 0;
    while (i < cfg.msgs) : (i += 1) {
        w.interface.writeAll(msg[0..cfg.size]) catch return;
        w.interface.flush() catch return;
        r.interface.readSliceAll(msg[0..cfg.size]) catch return;
    }
}

pub fn main(init: std.process.Init.Minimal) !void {
    const gpa = std.heap.smp_allocator;

    var arena_state = std.heap.ArenaAllocator.init(gpa);
    defer arena_state.deinit();
    const args = try init.args.toSlice(arena_state.allocator());

    var cfg: Config = .{};
    var iarg: usize = 1;
    while (iarg < args.len) : (iarg += 1) {
        const arg = args[iarg];
        if (std.mem.eql(u8, arg, "--zio") or std.mem.eql(u8, arg, "--zio-mt") or std.mem.eql(u8, arg, "--threaded")) {
            continue; // io backend selection, handled by IoBackend below
        }
        const split_pos = std.mem.findScalar(u8, arg, '=') orelse arg.len;
        const key = arg[0..split_pos];
        const target: *u64 = if (std.mem.eql(u8, key, "--conns"))
            &cfg.conns
        else if (std.mem.eql(u8, key, "--msgs"))
            &cfg.msgs
        else if (std.mem.eql(u8, key, "--size"))
            &cfg.size
        else {
            std.log.err("unknown argument '{s}'", .{arg});
            std.log.err("usage: tcp_echo [--zio | --zio-mt | --threaded] [--conns=N] [--msgs=N] [--size=N]", .{});
            std.process.exit(1);
        };
        const value = if (split_pos < arg.len) arg[split_pos + 1 ..] else blk: {
            iarg += 1;
            if (iarg >= args.len) {
                std.log.err("expected a value after {s}", .{key});
                std.process.exit(2);
            }
            break :blk args[iarg];
        };
        target.* = try std.fmt.parseUnsigned(u64, value, 10);
    }
    if (cfg.conns == 0 or cfg.msgs == 0 or cfg.size == 0 or cfg.size > max_msg_size) {
        std.log.err("--conns/--msgs must be at least 1, --size in [1, {d}]", .{max_msg_size});
        std.process.exit(2);
    }

    var backend: IoBackend = .none;
    try backend.init(gpa, init.args);
    defer backend.deinit();

    const io = backend.io();

    const start_time: std.Io.Clock.Timestamp = .now(io, .real);

    var ready: std.Io.Event = .unset;

    var group: std.Io.Group = .init;
    defer group.cancel(io);

    try group.concurrent(io, serverTask, .{ io, cfg, &ready });
    var i: u64 = 0;
    while (i < cfg.conns) : (i += 1) {
        try group.concurrent(io, clientTask, .{ io, cfg, &ready });
    }
    try group.await(io);

    const end_time = std.Io.Clock.Timestamp.now(io, .real);
    const duration = start_time.durationTo(end_time);
    const total_msgs = cfg.conns * cfg.msgs;
    const ns = duration.raw.toNanoseconds();
    const rate = if (ns > 0) total_msgs * 1_000_000_000 / @as(u64, @intCast(ns)) else 0;
    std.log.info("Duration: {f} ({d} msgs over {d} conns, size {d}, {d} msgs/s)", .{ duration.raw, total_msgs, cfg.conns, cfg.size, rate });
}

File: rust/src/bin/tcp_echo.rs
// Golden TCP benchmark on tokio: --conns concurrent loopback connections, each
// doing --msgs request/response round-trips of --size bytes against a
// per-connection echo handler. Presets:
//
//   defaults (1000 conns x 100 msgs x 64B)   many-connection throughput
//   --conns=1 --msgs=100000 --size=4096      single-connection latency chain
use std::time::Instant;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::{TcpListener, TcpStream};

const ADDR: &str = "127.0.0.1:18768";

fn parse_args() -> (usize, usize, usize, bool) {
    let (mut conns, mut msgs, mut size) = (1000usize, 100usize, 64usize);
    let mut single_thread = false;
    for arg in std::env::args().skip(1) {
        if arg == "--st" {
            single_thread = true;
            continue;
        }
        let (key, value) = match arg.split_once('=') {
            Some(kv) => kv,
            None => {
                eprintln!("usage: tcp_echo [--st] [--conns=N] [--msgs=N] [--size=N]");
                std::process::exit(1);
            }
        };
        let target = match key {
            "--conns" => &mut conns,
            "--msgs" => &mut msgs,
            "--size" => &mut size,
            _ => {
                eprintln!("unknown argument '{}'", arg);
                std::process::exit(1);
            }
        };
        *target = value.parse().expect("invalid number");
    }
    (conns, msgs, size, single_thread)
}

async fn echo_handler(mut stream: TcpStream, size: usize) {
    let mut msg = vec![0u8; size];
    loop {
        if stream.read_exact(&mut msg).await.is_err() {
            return;
        }
        if stream.write_all(&msg).await.is_err() {
            return;
        }
    }
}

async fn client(msgs: usize, size: usize) {
    let mut stream = TcpStream::connect(ADDR).await.expect("connect failed");
    stream.set_nodelay(true).ok();
    let mut msg = vec![0u8; size];
    for _ in 0..msgs {
        if stream.write_all(&msg).await.is_err() {
            return;
        }
        if stream.read_exact(&mut msg).await.is_err() {
            return;
        }
    }
}

async fn run(conns: usize, msgs: usize, size: usize) {
    let listener = TcpListener::bind(ADDR).await.expect("listen failed");

    let server = tokio::spawn(async move {
        let mut handlers = Vec::with_capacity(conns);
        for _ in 0..conns {
            let (stream, _) = match listener.accept().await {
                Ok(s) => s,
                Err(e) => {
                    eprintln!("server: accept failed: {}", e);
                    break;
                }
            };
            stream.set_nodelay(true).ok();
            handlers.push(tokio::spawn(echo_handler(stream, size)));
        }
        for h in handlers {
            h.await.ok();
        }
    });

    let mut clients = Vec::with_capacity(conns);
    for _ in 0..conns {
        clients.push(tokio::spawn(client(msgs, size)));
    }
    for c in clients {
        c.await.ok();
    }
    server.await.ok();
}

fn main() {
    let (conns, msgs, size, single_thread) = parse_args();
    let rt = if single_thread {
        tokio::runtime::Builder::new_current_thread()
            .enable_all()
            .build()
            .unwrap()
    } else {
        tokio::runtime::Builder::new_multi_thread()
            .enable_all()
            .build()
            .unwrap()
    };
    let start = Instant::now();
    rt.block_on(run(conns, msgs, size));
    let d = start.elapsed();
    let total = (conns * msgs) as f64;
    println!(
        "Duration: {:.3}ms ({} msgs over {} conns, size {}, {:.0} msgs/s)",
        d.as_secs_f64() * 1000.0,
        conns * msgs,
        conns,
        size,
        total / d.as_secs_f64()
    );
}
